src/main/java/nextstep/subway/line/domain/Sections.java

```diff
@@ -0,0 +1,129 @@
+package nextstep.subway.line.domain;
+
+import nextstep.subway.station.domain.Station;
+
+import javax.persistence.CascadeType;
+import javax.persistence.Embeddable;
+import javax.persistence.OneToMany;
+import java.util.*;
+
+@Embeddable
+public class Sections implements Iterable<Section> {
+
+    private static final int MIN_SECTIONS_SIZE = 1;
+
+    @OneToMany(mappedBy = "line", cascade = {CascadeType.PERSIST, CascadeType.MERGE}, orphanRemoval = true)
+    private List<Section> sections = new ArrayList<>();
+
+    protected Sections() {
+    }
+
+    public void add(Section section) {
+        if (isEmpty()) {
+            sections.add(section);
+            return;
+        }
+        if (!contains(section.getUpStation()) && !contains(section.getDownStation())) {
+            throw new IllegalArgumentException("구간 추가를 위해서는 기존 구간과의 연결점이 필요합니다. 역 정보를 확인해주세요.");
+        }
+        mergeByUpStation(section);
+        mergeByDownStation(section);
+        sections.add(section);
+    }
+
+    private void mergeByUpStation(Section section) {
+        sections.stream()
+                .filter(it -> it.getUpStation() == section.getUpStation())
+                .findFirst()
+                .ifPresent(it -> it.updateUpStation(section.getDownStation(), section.getDistance()));
+    }
+
+    private void mergeByDownStation(Section section) {
+        sections.stream()
+                .filter(it -> it.getDownStation() == section.getDownStation())
+                .findFirst()
+                .ifPresent(it -> it.updateDownStation(section.getUpStation(), section.getDistance()));
+    }
+
+    public void remove(Line line, Station station) {
+        validateRemovable(station);
+
+        Optional<Section> upStationMatching = sections.stream()
+                .filter(it -> it.getUpStation() == station)
+                .findFirst();
+        Optional<Section> downStationMatching = sections.stream()
+                .filter(it -> it.getDownStation() == station)
+                .findFirst();
+
+        if (upStationMatching.isPresent() && downStationMatching.isPresent()) {
+            Station newUpStation = downStationMatching.get().getUpStation();
+            Station newDownStation = upStationMatching.get().getDownStation();
+            int newDistance = upStationMatching.get().getDistance() + downStationMatching.get().getDistance();
+            sections.add(new Section(line, newUpStation, newDownStation, newDistance));
+        }
```

Comment:
upStationMatcing과 downStationMatching이 모두 존재하는 경우는 어떤 경우일까요?

그리고 그 경우 어떤 행동을 수행해야 할까요?

지금은 위 맥락 파악을 위해서 지하철과 관련된 도메인의 파악 및 내부 코드를 자세히 살펴보아야 할 것 같아요.

메소드 분리를 통해 의도를 드러내주시고, 내부에서 어떤 행위를 하는지도 나타내주시면 좋을 것 같습니다.

Reply:
네 규남님~! 저도 이 로직에 대해 고민이 좀 많았습니다😅

먼저 저 if 조건문에서 표현되어야 할 의도는 제거하고자 하는 station 이 종착역인지 에 대한 여부 라고 생각했습니다.
그런데 station 에 대한 종착역 여부를 판단하기 위해 !isEdge(station) 또는 isMiddleStation(station) 과 같은 메소드를 추출하자니 메소드 내부에서 다시 findUpSection, findDownSection 메소드를 호출해야 하겠더라구요ㅠㅠ

그렇다면 isMiddle(upStationMatching, downStationMatching) 과 같이 두 Optional 인스턴스를 넘겨서 판단을 해볼까 하는 생각도 했으나, 이렇게 하면 메소드의 시그니처가 isMiddle(Optional<Section>, Optional<Section>) 로 정의가 되어, 이런 형태는 부자연스럽겠다는 판단을 했습니다.

결국에는 변수명을 변경하여 upSection.isPresent() && downSection.isPresent() 조건으로
'상행구간과 하행구간이 존재한다' -> 중간에 위치한 역이다 라는 의도를 표현하고,
조건문 내부의 행동은 addMergeSections 메소드로 추출하여 두 구간을 통합한 새로운 구간을 추가하도록 했는데요!
IntelliJ 에서 Optional 인스턴스를 파라미터로 넘기는 코드에 대해 warning 메시지가 나오고 있습니다.

제가 Optional 을 활용해 본 경험이 많지 않아서, 이러한 코드가 자연스러운지에 대한 판단이 잘 서지가 않는 것 같습니다.
Optional 인스턴스를 파라미터로 넘기는 모양이 안티패턴일까요?😅 지양해야 할 부분이라면 remove 프로세스의 구조에 대해서 다시 고민해 보아야 할 것 같습니다😭

Reply:
Optional의 경우 파라미터로 전달되는 부분은 권장되지 않습니다.

아래 문서를 한번 읽어보시면 좋을 것 같아요

## Item 16: Do Not Use Optional in Methods Arguments

```
Don't force call sites to create Optionals. Do not use Optional as fields or in setters and constructors' arguments.

UsingOptionalin methods arguments is another common mistake. This would lead to code that is unnecessarily complicated. Take responsibility for checking arguments instead of force call sites to create Optionals. This practice clutters the code and may cause dependence. In time, you will use it everywhere. Keep in mind thatOptionalis just another object (a container) and is not cheap — it consumes 4x the memory of a bare reference! But you may also like to check this out and decide, depending on the case.
```

https://dzone.com/articles/using-optional-correctly-is-not-optional